src/v/kafka/server/handlers/fetch.cc

```diff
@@ -106,8 +125,22 @@ static ss::future<read_result> read_from_partition(
     std::unique_ptr<iobuf> data;
     std::vector<cluster::rm_stm::tx_range> aborted_transactions;
     try {
-        auto result = co_await rdr.reader.consume(
-          kafka_batch_serializer(), deadline ? *deadline : model::no_timeout);
+        const auto ts_deadline = deadline.value_or(model::no_timeout);
+        const auto failsafe_deadline = failsafe_deadline_from(ts_deadline);
+        auto consume_fut = rdr.reader.consume(
+          kafka_batch_serializer(), ts_deadline);
+        auto result
+          = co_await ss::with_timeout(failsafe_deadline, std::move(consume_fut))
+              .handle_exception_type(
```

Comment:
If this times out, what guarantees that the consuming fiber doesn't continue trying to access `rdr` after it's destructed?

Reply:
If consume times out, we catch the exception and store it in a variable, then we wait for 

```
co_await std::move(rdr.reader).release()->finally();
```

before rethrowing the exception.

Reply:
I think this safely closes the reader, but please let me know if you think there could still be issues?